### DISCUSSION TITLE
Enforcing strict back pressure

### DISCUSSION THREAD

**Discussion by paulfariello** [rabbitmq/rabbitmq-server#3026] [Questions]
I'm looking for a way to force strong back pressure to producers.

I'm trying to use flow throttling and avoid paging out. Following https://www.rabbitmq.com/memory.html, it seems that setting a `vm_memory_high_watermark_paging_ratio` to > `1` should be enough to avoid paging and using `vm_memory_high_watermark.absolute` should allow to force stopping producer connection when the limit is reached.

Unfortunately, trying to set `vm_memory_high_watermark_paging_ratio` to `1` fails with `vm_memory_high_watermark_paging_ratio invalid, Float is not between 0 and 1`.

It seems that this check is here since a long time. https://github.com/rabbitmq/rabbitmq-server/blob/f49bb000d19ac30b3ef42d0e20d7d360fd3c29fd/deps/rabbit/priv/schema/rabbit.schema#L774

Is there another way to enforce a strict back pressure. Meaning, if memory threshold is reached, producer connection is totally frozen?

Btw, congrats for the work done on RabbitMQ, it's a great piece of software.

**Comment by michaelklishin** [comment]
RabbitMQ will stop reading from the socket of connections that send at least one message (one of the `basic.publish` frames) when a memory alarm is in effect. That's the "strictest" back pressure you have.

Watermark paging ratio controls to what extent classic queues will delay paging in memory messages. Setting it to 1 or higher is dangerous. While setting validation can be tweaked, we usually see people who want to do the opposite: make memory usage more predictable, and a higher watermark for CQs will do the opposite of that.

The back pressure mechanism is "strict" by your definition. I'm not sure what you are trying to achieve here. Quorum queues have
a more predictable peak memory usage controls but they by definition do not try to delay disk operations: that's the whole point of Raft and its data safety promises.

More or less precise resource caps in highly concurrent distributed system are only possible if you allocate all memory directly and RabbitMQ by definition does not: the Erlang runtime does. Or the kernel cap do it (cgroups) but its reaction is not something you typically want in a stateful data service.
